src/Microsoft.OData.Core/Json/JsonReaderExtensions.cs

```diff
@@ -237,76 +243,134 @@ internal static void SkipValue(this IJsonReader jsonReader)
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="jsonReader"/> is <see langword="null"/></exception>
+        /// <exception cref="ODataException">
+        /// Thrown if the <see cref="IJsonReader.NodeType"/> of <paramref name="jsonReader"/> is not <see cref="JsonNodeType.StartArray"/> or the JSON that
+        /// <paramref name="jsonReader"/> is reading is not well-formed
+        /// </exception>
+        internal static IEnumerable<ODataUntypedValue> ReadUntypedCollectionValues(this IJsonReader jsonReader)
+        {
+            if (jsonReader == null)
+            {
+                throw new ArgumentNullException(nameof(jsonReader));
+            }
 
-                        default:
-                            Debug.Assert(
-                                jsonReader.NodeType != JsonNodeType.EndOfInput,
-                                "We should not have reached end of input, since the scopes should be well formed. Otherwise JsonReader should have failed by now.");
-                            break;
-                    }
-                }
-                while (jsonReader.Read() && depth > 0);
+            return jsonReader.ReadUntypedCollectionValuesIterator();
+        }
 
-                if (depth > 0)
-                {
-                    // Not all open scopes were closed:
-                    // "Invalid JSON. Unexpected end of input was found in JSON content. Not all object and array scopes were closed."
-                    throw JsonReaderExtensions.CreateException(Strings.JsonReader_EndOfInputWithOpenScope);
-                }
+        /// <summary>
+        /// Reads the elements in a JSON array and parses them as <see cref="ODataUntypedValue"/>s
+        /// </summary>
+        /// <param name="jsonReader">
+        /// The <see cref="IJsonReader"/> that the collection of untyped values should be read from; assumed to not be <see langword="null"/>
+        /// </param>
+        /// <returns>The untyped values in the JSON array in <paramref name="jsonReader"/></returns>
+        /// <exception cref="ODataException">
+        /// Thrown if the <see cref="IJsonReader.NodeType"/> of <paramref name="jsonReader"/> is not <see cref="JsonNodeType.StartArray"/> or the JSON that
+        /// <paramref name="jsonReader"/> is reading is not well-formed
+        /// </exception>
+        private static IEnumerable<ODataUntypedValue> ReadUntypedCollectionValuesIterator(this IJsonReader jsonReader)
+        {
+            jsonReader.ReadStartArray();
 
-                jsonWriter.Flush();
+            while (jsonReader.NodeType != JsonNodeType.EndArray)
+            {
+                yield return jsonReader.ReadAsUntypedOrNullValueImpl();
             }
+
+            jsonReader.ReadEndArray();
         }
 
         internal static ODataValue ReadAsUntypedOrNullValue(this IJsonReader jsonReader)
+        {
+            return jsonReader.ReadAsUntypedOrNullValueImpl();

```

Comment:
I see. As for the name, not sure. I've seen `Internal` suffix being used in some parts of the codebase and `Implementation` in other parts (especially the writer and serializers). So, for consistence I would go for `Implementation` instead of `Impl`.